aws-datastore/src/main/java/com/amplifyframework/datastore/storage/sqlite/ModelTreeHelper.java

```diff
@@ -0,0 +1,140 @@
+
+import com.amplifyframework.core.model.Model;
+import com.amplifyframework.core.model.ModelAssociation;
+import com.amplifyframework.core.model.ModelSchema;
+import com.amplifyframework.core.model.ModelSchemaRegistry;
+import com.amplifyframework.core.model.query.QueryOptions;
+import com.amplifyframework.core.model.query.Where;
+import com.amplifyframework.core.model.query.predicate.QueryField;
+import com.amplifyframework.core.model.query.predicate.QueryPredicate;
+import com.amplifyframework.core.model.query.predicate.QueryPredicateOperation;
+import com.amplifyframework.core.model.query.predicate.QueryPredicates;
+import com.amplifyframework.datastore.DataStoreException;
+import com.amplifyframework.datastore.storage.sqlite.adapter.SQLiteTable;
+import com.amplifyframework.util.Empty;
+
+import java.util.Collection;
+import java.util.HashSet;
+import java.util.LinkedHashMap;
+import java.util.Map;
+import java.util.Set;
+
+/**
+ * Utility class to help traverse a tree of models by relationship.
+ */
+final class ModelTreeHelper {
+    private final ModelSchemaRegistry registry;
+    private final SQLiteStorageAdapter storage;
+
+    /**
+     * Constructs a model family tree traversing utility.
+     * @param registry model registry to search schema from
+     * @param storage SQLite storage engine
+     */
+    ModelTreeHelper(ModelSchemaRegistry registry,
+                    SQLiteStorageAdapter storage) {
+        this.registry = registry;
+        this.storage = storage;
+    }
+
+    /**
+     * Returns a map of descendants of a set of models (of same type).
+     * A model is a child of its parent if it uses its parent's ID as foreign key.
+     * @param root Collection of models to query its descendants of.
+     * @return Map of descendants keyed by model schema. The value contains a set of
+     *          descendants' IDs for that model type.
+     */
+    <T extends Model> Map<ModelSchema, Set<String>> descendantsOf(Collection<T> root) {
+        if (Empty.check(root)) {
+            throw new IllegalArgumentException("Cannot traverse tree from an empty root.");
+        }
+        Map<ModelSchema, Set<String>> descendants = new LinkedHashMap<>();
+        ModelSchema rootSchema = registry.getModelSchemaForModelInstance(root.iterator().next());
+        Set<String> rootIds = new HashSet<>();
+        for (T model : root) {
+            rootIds.add(model.getId());
+        }
+        recurseTree(descendants, rootSchema, rootIds);
+        return descendants;
+    }
+
+    private void recurseTree(
+            Map<ModelSchema, Set<String>> tree,
+            ModelSchema modelSchema,
+            Collection<String> parentIds
+    ) {
+        SQLiteTable parentTable = SQLiteTable.fromSchema(modelSchema);
+        for (ModelAssociation association : modelSchema.getAssociations().values()) {
+            switch (association.getName()) {
+                case "HasOne":
+                case "HasMany":
```

Comment:
Can we get these relationship strings into an enum somewhere?

Also, the contents of the case statements looks like a good candidate for some smaller helper method(s). That will also help to reduce the nesting.

Reply:
The content of the case statement _does_ feel long, but I believe it may be even more difficult to follow if they are comparmentalized into helper methods. With how many local variables there are being used in each step, it will become messier to try to move them away from the main body.

Reply:
With regards to turning `ModelAssociation#name` variable to enum, I think it's a good idea, but I'm afraid that it may potentially break some customers that are not relying on codegened models (custom models) in their project.